backend/apps/ifc_validation/tasks.py

```diff
@@ -609,27 +609,43 @@ def is_schema_error(line):
                     observed=None
                 )
             else:
+                outcomes_to_save = list()
+                outcomes_instances_to_save = list()
+
                 for line in output:
                     message = json.loads(line)
                     model.status_schema = Model.Status.INVALID
-                    outcome = task.outcomes.create(
+                    outcome = ValidationOutcome(
                         severity=ValidationOutcome.OutcomeSeverity.ERROR,
                         outcome_code=ValidationOutcome.ValidationOutcomeCode.SCHEMA_ERROR,
                         observed=message['message'],
                         feature=json.dumps({
                             'type': message['type'] if 'type' in message else None,
                             'attribute': message['attribute'] if 'attribute' in message else None
-                        }),
+                        })
                     )
+                    outcome.validation_task = task
+                    outcomes_to_save.append(outcome)
 
                     if 'instance' in message and message['instance'] is not None and 'id' in message['instance'] and 'type' in message['instance']:
-                        instance, _ = model.instances.get_or_create(
+                        instance = ModelInstance(
```

Comment:
I guess there is a small change in behaviour here where before get_or_create() would not create duplicate values based on the stepfile_id,ifc_type provided, but now now insertion happens in a list with duplicates and only after insertion and after retrieval from DB these are reorganized in a dict, leaving the duplicates inaccessible from python, but still there in the DB. I guess nobody would notice, but putting these orphans in the db is still not very nice.

Maybe consider

```
outcomes_instances_to_save = dict()
...
if 'instance' in message and message['instance'] is not None and 'id' in message['instance'] and 'type' in message['instance'] \
** and message['instance']['id'] not in outcomes_instances_to_save **:
   instance = ModelInstance(...)
   outcomes_instances_to_save[message['instance']['id']] = instance
...
model_instances = dict(zip(outcomes_instances_to_save.keys(), ModelInstance.objects.bulk_create(list(outcomes_instances_to_save.values())))))))
```

Because there don't seem to be any unique keys on ModelInstance. Or did I miss them?

Reply:
I actually had an update to ModelInstance to enforce a unique constraint for this.
This + line below would enforce uniqueness in DB.

Line 638:
`ModelInstance.objects.bulk_create(outcomes_instances_to_save, ignore_conflicts=True)`